src/diart/models.py

```diff
@@ -143,33 +162,25 @@ def from_pyannote(
         assert _has_pyannote, "No pyannote.audio installation found"
         return PyannoteEmbeddingModel(model, use_hf_token)
 
-    @abstractmethod
-    def forward(
-        self, waveform: torch.Tensor, weights: Optional[torch.Tensor] = None
-    ) -> torch.Tensor:
-        """
-        Forward pass of an embedding model with optional weights.
-
-        Parameters
-        ----------
-        waveform: torch.Tensor, shape (batch, channels, samples)
-        weights: Optional[torch.Tensor], shape (batch, frames)
-            Temporal weights for each sample in the batch. Defaults to no weights.
-
-        Returns
-        -------
-        speaker_embeddings: torch.Tensor, shape (batch, embedding_dim)
-        """
-        pass
-
 
 class PyannoteEmbeddingModel(EmbeddingModel):
     def __init__(self, model_info, hf_token: Union[Text, bool, None] = True):
         super().__init__(PyannoteLoader(model_info, hf_token))
 
-    def forward(
-        self,
-        waveform: torch.Tensor,
-        weights: Optional[torch.Tensor] = None,
+    def __call__(
+        self, waveform: torch.Tensor, weights: Optional[torch.Tensor] = None
     ) -> torch.Tensor:
-        return self.model(waveform, weights=weights)
+        if not isinstance(self.model, WeSpeakerPretrainedSpeakerEmbedding):
+            return super().__call__(waveform, weights)
+        else:
+            self.load()
+            # Normalize weights
+            if weights is not None:
+                weights -= weights.min(dim=1, keepdim=True).values
+                weights /= weights.max(dim=1, keepdim=True).values
+                weights.nan_to_num_(0.0)
+                # Move to cpu for numpy conversion
+                weights = weights.to("cpu")
+            # Move to cpu for numpy conversion
+            waveform = waveform.to("cpu")
+            return torch.from_numpy(self.model(waveform, weights))
```

Comment:
do waveform and weights have to be numpy? have you checked that this works with `pyannote/embedding`? Also please check that this way we can run both models on GPU.

If you don't have a GPU, I can check this myself during testing. Please let me know.

Also, please use `super().__call__(waveform, weights)` here as well so you don't have to call `self.load()` manually

Reply:
Yeah, sadly they have to be. It will be executed on the GPU, but [this](https://github.com/pyannote/pyannote-audio/blob/0b45103cb228a81a9d9d776cca92694cb30ddb41/pyannote/audio/pipelines/speaker_verification.py#L609) `.numpy()` requires the `waveform.to("cpu")`
Additionally [this](https://github.com/pyannote/pyannote-audio/blob/0b45103cb228a81a9d9d776cca92694cb30ddb41/pyannote/audio/pipelines/speaker_verification.py#L603) requires the `weights.to("cpu")` as we otherwise fail with
```python
RuntimeError: indices should be either on cpu or on the same device as the indexed tensor (cpu)
```

I can verify that the GPU is utilized, you can see the ONNX provider set with our call to `.to("cuda)` because of this [definition](https://github.com/pyannote/pyannote-audio/blob/0b45103cb228a81a9d9d776cca92694cb30ddb41/pyannote/audio/pipelines/speaker_verification.py#L441)

Nonetheless, the tensors have to be converted to numpy for onnx use and as such the errors pointed out above exist if they are not moved to the cpu beforehand.

Reply:
Do all types of embedding models use onnx? Maybe @hbredin can give us a hint here.

If we absolutely need this `to("cpu")`, we could create a wrapper around `PretrainedEmbeddingModel`, let's say `PipelineInputFormatter`.

This way `PyannoteLoader` would return a `PipelineInputFormatter` instead of a `PretrainedEmbeddingModel`:

```python
class PipelineInputFormatter:
    def __init__(self, model: PretrainedEmbeddingModel):
        self.model = model

    def __call__(self, audio, masks) -> np.ndarray:
        return self.model(audio.cpu().numpy(), masks.cpu().numpy())
```

I don't really like the name though, I'm open to suggestions :)

Reply:
It may also be a matter of needing to specify the device upon instantiation, contrary to `nn.Module`. In that case we may want to refactor diart models to work in the same way

Reply:
> Do all types of embedding models use onnx? 

No. Only `WeSpeaker` ones.

Reply:
Ok I just took a look at the links sent by @sorgfresser and it does look like a pyannote bug.
@hbredin I opened a PR to fix this: https://github.com/pyannote/pyannote-audio/pull/1518
Could you take a look?

Also, if it's not too much to ask, it would be great to have a release with this fix so we don't have to do any weird workarounds here